### DISCUSSION TITLE
How to trim strings before validating ?

### DISCUSSION THREAD

**Discussion by Ahmad-Fathy** [filamentphp/filament#6868] [General]
I want to make a unique record in a table
But it saves values like "New book", "       New  book" and  "New       book        " as a deffernt values
what i want is trim any multibles spaces to one space so the validate fire up as a duplicated value
i try to make a middleware that do this and add it in config/filament,php but it not working

**Comment by francoism90** [comment]
@Ahmad-Fathy Did you ever found a solution? The middleware doesn't also work for me.

I'm thinking of creating a PR to add `->trim()` to TextInput fields.

**Reply by leandrocfe** [reply]
You can use [lifecycle-hooks](https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks)
CreatePage / EditPage
```php
protected function beforeValidate(): void
{
    $this->data['title'] = trim($this->data['title']);
}
```

**Reply by francoism90** [reply]
@leandrocfe That doesn't work, as the data can differ (for me when using Spatie translations).

**Reply by Ahmad-Fathy** [reply]
I make a custom validation rule to trime more thane one space at any part of data

**Reply by francoism90** [reply]
@Ahmad-Fathy Could you please share your solution?

**Reply by Ahmad-Fathy** [reply]
This code to validate that input has no more one space .. it's about Arabic language, you can customise it as you want

```
class CheckSpacesRule implements ValidationRule
{
  public function validate(string $attribute, mixed $value, Closure $fail): void
  {
    if (preg_match('/^([ء-ي]+\s)*[ء-ي]+$/u', $value) == 0) {
      $fail('validation.check_spaces')->translate();
    }
  }
}
```
